enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,216 @@
+be caused by these factors.
+
+If an update requires downloading new container images on a relatively slow
+network, it may take more than the default wait period to run this operation.
+The mitigation would be automatic doubling of the wait period on subsequent
+boots. If this is still not enough, system administrators should set the base
+wait period setting to a higher value to start with.
+
+If any external connectivity is required for upgrades to complete successfully,
+intermittent network outage may result in a rollback to the previous version.
+As a mitigation, system administrators should reinstall the upgrade when network
+is available for the Edge Devices.
+
+To further mitigate potential problems related to slow or unstable networks,
+system administrators may choose to bundle the upgraded container images in the
+`ostree` software upgrade. In this case, when the upgrade is installed, no
+more container images will need to be downloaded over the network.
+
+Some 3rd party workloads may become operational before the upgrade is declared
+valid and potentially create or update data on the device. If a rollback is
+performed subsequentially, there is a risk of data loss because the file system
+is reverted to its state before the upgrade. To mitigate this problem, 3rd party
+workloads should never become operational until a boot is declared successful.
+
+```bash
+$ sudo grub2-editenv - list | grep ^boot_success
+boot_success=1
+```
+
+### Drawbacks
+
+The technical implementation of upgrade validation is rather simple as it relies
+on the existing `greenboot` and `rpm-ostree` facilities. However, end users may
+find it annoying to have their Edge Devices rebooted few times until the upgrade
+is either successful or rolled back.
+
+It is possible that certain 3rd party workloads become operational before the
+upgrade is declared successful. This situation may further hurt user experience
+if their Edge Devices are rebooted due to MicroShift core functionality
+validation not passing.
+
+## Design Details
+
+### Open Questions
+
+* Do we need more validations to be implemented on upgrade?
+* Do we need to implement internal validations in MicroShift?
+* Do we need to add special log messages to be parsed by the validation?
+* Is there a need to implement post-boot scripts in case of success or failure? 
+
+### Upgrade / Downgrade Strategy
+
+The validation scripts will be delivered as part of the MicroShift core RPM
+packages. Any problems caused by the ugraded software, scripts or system
+configuration will be rolled back by `greenboot` when validations do not pass.
+
+## Alternatives
+
+The basic assumption is that `greenboot` must be used for implementing health
+checks on `rpm-ostree` based systems. It is possible to implement
+[Health Checks with systemd services](https://github.com/fedora-iot/greenboot#health-checks-with-systemd-services)
+if parts of the procedure should be executed taking into account unit ordering.
```

Comment:
Does greenboot honor those systemd checks? Should we be using that integration?

Reply:
It is just another way to start health check scripts, taking into account systemd dependencies. I do not think it's necessary at the moment, but we can always implement it if needed.